### ISSUE TITLE
Problems with -F (send to / receive from) file

### ISSUE THREAD

**Issue by shd2015** [esnet/iperf#301]
Hi,
I am using iperf3 on Debian Linux OS. I have two machines and each has one 10Gbps ethernet port. The 10G ports are connected directly via an optical fiber.
I am trying to send a data file from client to server using iperf3 with -F option using UDP transport. On the server end, I want to write the received data to a file. 
The command I am using at the client is:
iperf3 -c <ip address of the server> -u -t30 -l14K -A0 -b10G -F 4Kfile.mp4 -P1 -p5001

The command I am using at the server is:
iperf3 -s -F 4KFileRx.mp4 -p5001

Issues I see are:
1. When I use the -F option at the server, I see 100% data loss and some garbage gets written into the file. Question 1: What is the mistake I am doing and why it is discarding all the data ?
1. If I remove -F option at the server, then the client is able to read from the file specified and sends the data until all the data is sent. I see ~2 to ~3% datagram loss and data rate of ~4.5Gbps. 
   Question 2: The data transfer stops as soon as all the data is read from the file. How do I make it continuous and stop it only after 30 sec as specified in the command ?
   Question 3: I want to increase the data rate to ~10Gbps. If I use -P2 option, does iperf duplicate the data on both the streams or does it split the data such that half of the file goes on one stream and the other half on the other stream ? From my observation, it looked like it is duplicating the data on both the streams.

I would appreciate if anyone can share some thoughts on my questions. Thanks in advance.

Regards,
SHD

**Comment by bltierney** [comment]
There definitely seems to be a bug here. 

Even TCP does not work right. Even asking for -t200, the transfer ends after 10 seconds:

>iperf3 -p5400 -c nettest.lbl.gov -t200 -F testfile
Connecting to host nettest.lbl.gov, port 5400
[  4] local 198.128.151.25 port 60788 connected to 131.243.24.11 port 5400
[ ID] Interval           Transfer     Bandwidth       Retr  Cwnd
[  4]   0.00-1.00   sec  1.03 MBytes  8.68 Mbits/sec    0   36.8 KBytes
[  4]   1.00-2.00   sec  1.26 MBytes  10.6 Mbits/sec    0   55.1 KBytes
[  4]   2.00-3.00   sec  1.93 MBytes  16.2 Mbits/sec    0   70.7 KBytes
[  4]   3.00-4.00   sec  3.23 MBytes  27.1 Mbits/sec    0    103 KBytes
[  4]   4.00-5.00   sec  3.67 MBytes  30.8 Mbits/sec    0    134 KBytes
[  4]   5.00-6.00   sec  4.41 MBytes  37.0 Mbits/sec    0    154 KBytes
[  4]   6.00-7.00   sec  3.36 MBytes  28.1 Mbits/sec    0    156 KBytes
[  4]   7.00-8.00   sec  3.79 MBytes  31.8 Mbits/sec    0    157 KBytes
[  4]   8.00-9.00   sec  4.47 MBytes  37.5 Mbits/sec    0    161 KBytes
[  4]   9.00-9.62   sec  2.36 MBytes  32.2 Mbits/sec    0    163 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-9.62   sec  29.5 MBytes  25.7 Mbits/sec    0             sender
        Sent 29.5 MByte /  124 MByte (23%) of testfile
[  4]   0.00-9.62   sec  28.2 MBytes  24.6 Mbits/sec                  receiver

only 28M of the 124M file were transferred.

And for UDP, it ends even faster, claiming it sent the entire file, but it did not:

>iperf3 -p5400 -c nettest.lbl.gov -t200 -F testfile -u -b500M
Connecting to host nettest.lbl.gov, port 5400
[  4] local 198.128.151.25 port 54325 connected to 131.243.24.11 port 5400
[ ID] Interval           Transfer     Bandwidth       Total Datagrams
[  4]   0.00-1.00   sec  41.9 MBytes   352 Mbits/sec  30346
[  4]   1.00-2.00   sec  76.0 MBytes   638 Mbits/sec  55071
[  4]   2.00-2.08   sec  5.94 MBytes   626 Mbits/sec  4302
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-2.08   sec   124 MBytes   500 Mbits/sec  37.582 ms  3756/3808 (99%)
[  4] Sent 3808 datagrams
        Sent  124 MByte /  124 MByte (100%) of testfile

iperf Done.

But the file on the receiver the file was only 88K, not 124M.

Perhaps all that is needed is to flush the file to disk before ending the test? Perhaps there are other issues as well.

I suggest that by default, including -F should disable the default test duration of 10seconds, and it should just transfer the entire file and then exit. However, if there is a -t flag specified, the test should transfer the file until time 't', and then exit.

**Comment by bltierney** [comment]
and regarding question 3 above, if you use -F with -P > 1, the resulting file will likely be garbage. iperf3 is not a parallel data transfer tool. The reason for the -F flag is to help determine if the network or the disk is the bottleneck, as described here: https://fasterdata.es.net/performance-testing/network-troubleshooting-tools/iperf/disk-testing-using-iperf/

**Comment by bmah888** [comment]
Something weird is happening.  In my testing, I'm seeing the sender/client send off all the bytes of the file (subject to `-t`), and the receiver/server sees to be think it's received them, but the bytes aren't all hitting the disk.  I'm not sure what's up with that, because I've never used this mode of operation (`-F` on the server) myself.

OP, I don't recommend doing `-l14K` unless the NICs you are using between your two machines can actually support a UDP datagram of that size.  Otherwise you're like to see UDP fragmentation and possibly higher packet loss.

**Comment by bmah888** [comment]
OK.  I have found at least four issues with `iperf -F`.  Some of them are outright bugs, and others could be called "non-intuitive behavior".  I found all of these with the tip of master code (with some diagnostics code added) running on a FreeBSD 11 VM:

1.  When sending from a file, the sending side (whether it's the client or server) reads the file in chunks equal to the block size (`-l` parameter).  The last chunk of the file might not be exactly a multiple of the block size, but we try to send an entire block size anyway.  This might result in some amount of garbage being sent after the end of the file.

2.  Primarily a problem for TCP:  Each iteration through the sending loop reads a chunk of the file (as described above) into a buffer and then sends it using `Nwrite()`.  But it's possible that the send operation doesn't actually accept all of the data, for example because the sending socket buffer is full.  Regardless, on the next iteration through the sending loop, iperf3 reads another chunk of the file, regardless of whether the previous chunk was completely sent.  This can result in random chunks of the file not being sent on the network at all.  This can be somewhat mitigated by using the `--bandwidth` option to force iperf3 to slow down.

3.  On the receiver side, there's a spurious (and wrong) message printed about "Sent ... Bytes".  We need to not do that.

4.  This is purely a UDP problem.  When doing `--udp`, iperf3 will read chunks of the file and send them as UDP datagrams.  But the first bytes of the UDP payload of each datagram are overwritten by a header that iperf3 uses for sequence numbers and timing information.  So even if there is no loss or corruption in the network, the file on the receiving end won't match the contents on the sending end.

As I'm thinking about this, I think that items 2 and 3 above are bugs, while the others deserve at least a mention in the documentation, primarily that even in the case of a successful transfer, the source and destination files might not be identical.

**Comment by bltierney** [comment]
I agree that 2 and 3 should be fixed, and maybe not worry about 1 and 4, and iperf3 should not be used as a file transfer tool (tho the behavior should certainly be documented).

**Comment by bmah888** [comment]
Addendum to point 3 above:  The human-readable message on the receiver side looks like this:
```
        Sent 0.00 Byte / 33.4 MByte (0%) of fubar
```
The "sent" is technically correct because this is the receiver side, and it technically didn't send any bytes.  Of course that's totally useless because we should in this case be printing the amount of data "received".  Unless something bizarro happens that should be the same as the file length.

On the receiver side this data gets written to the JSON as well...this structure gets written per-stream in the ending statistics:
```
				"diskfile":	{
					"sent":	0,
					"size":	32783880,
					"percent":	0,
					"filename":	"fubar"
				},
```
I'd like to know if anything depends on this part of the JSON structure.  I suppose we could make it print appropriate things based on whether it's the sender or receiver, but if we do any changes to this structure it might confuse other programs that use it.

Addressing a comment earlier about `-P`:

5.  We should probably disallow the combination of `-P` and `-F`, although the world won't end if someone tries it (they'll probably just get some weird result).

**Comment by bltierney** [comment]
I'm not aware of anything that uses the 'diskfile' stuff in the JSON.

**Comment by bmah888** [comment]
Fixed by pull request #588.

**Comment by rdp** [comment]
I wish iperf3 could be used as a file transfer tool since it is faster than basically any other options here: https://unix.stackexchange.com/a/471951/8337 :D
